### PR TITLE
vz: add adaptive memory ballooning for macOS Virtualization Framework

### LINKED ISSUE #4220: Support memory ballooning in VMs
### Description

I have a VM which uses approximately half of its allocated memory most of the time but sometimes spikes up near the allocated memory limit. To ensure uninterrupted execution of processes inside the VM, I have to allocate much more memory than the VM needs most of the time.

This could be improved if Lima VMs released unused memory automatically without user intervention.

Lima doesn't support memory ballooning currently. Apple docs [indicate](https://developer.apple.com/documentation/virtualization/vzvirtualmachineconfiguration/memoryballoondevices) that VZ supports this but I'm not sure of the scope of the changes required.

Relevant discussion: https://github.com/lima-vm/lima/discussions/2720

### PR DESCRIPTION
**Pull request by jwehrlich**
## Summary

Add adaptive memory ballooning for macOS Virtualization Framework (VZ) VMs, allowing Lima to dynamically adjust guest memory based on actual pressure rather than requiring peak allocation upfront.

Closes #4220

## Commits

This PR contains 4 logically grouped commits that build on each other:

1. **limatype,limayaml: add MemoryBalloon config and validation** — Adds the `MemoryBalloon` struct with 4 user-facing fields (`enabled`, `min`, `idleTarget`, `cooldown`), validation, and documentation.

2. **guestagent: add memory metrics collection** — Adds `GetMemoryMetrics` gRPC endpoint collecting `/proc/meminfo`, `/proc/pressure/memory`, `/proc/vmstat`, and container cgroup stats. Stateful collector tracks deltas for per-second rates.

3. **hostagent: add balloon controller state machine** — Six-state machine (Bootstrap → LearningDescend → Steady → OOMRecovery → CircuitBreaker → AgentFailure) with PSI-driven grow/shrink decisions, shrink guards, learned floor persistence, and host pressure integration. ~1300 lines of tests.

4. **vz: wire balloon controller to VZ driver** — Implements `Ballooner` interface in VZ driver, wires controller into hostagent polling loop, fills sensible defaults (min=25%, idleTarget=33%, cooldown=30s).

## Configuration

```yaml
vmOpts:
  vz:
    memoryBalloon:
      enabled: true
      min: "2GiB"
      idleTarget: "3GiB"
      cooldown: "30s"
```

All advanced tuning (thresholds, step sizes, guards) is hardcoded with sensible defaults to keep the user-facing API minimal.

## Key Design Decisions

- **4 public config fields only** — Advanced knobs (growStepPercent, shrinkStepPercent, pressure thresholds) are internal constants, not user-facing config.
- **cgroupfs for container metrics** — Reads cgroup2 `cpu.stat`/`io.stat` directly instead of depending on Docker API.
- **Learned floor** — Persists minimum observed working memory to `~/.lima/<instance>/balloon_floor` across restarts.
- **Host pressure awareness** — Monitors macOS `memory_pressure` to avoid shrinking guest when host is stressed.

## Testing

- Unit tests for all new packages: validation, metrics parsing, controller state machine, host pressure, learned floor
- ~2700 lines of test code covering states, transitions, edge cases, and guard conditions
- Every commit builds, lints, and tests independently

### CODE REVIEW

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
This amount of PR is not reviewable.

Please split the PR into multiple ones.
Notably auto-pause stuff does not seem directly related to memory optimization and does not need to be overloaded in this PR?

**Comment by jwehrlich** [comment]
> This amount of PR is not reviewable.
> 
> Please split the PR into multiple ones. Notably auto-pause stuff does not seem directly related to memory optimization and does not need to be overloaded in this PR?

@AkihiroSuda, I was thinking the same thing, but also wanted to get something in to do some initial validations to see if any tests failed/etc. I'll try to break this up into multiple PRs over the next week.

File: pkg/guestagent/api/guestservice.proto [hunk 1]
```diff
@@ -53,3 +56,36 @@ message TimeSyncResponse {
   int64 drift_ms = 2;
   string error = 3;
 }
+
+// MemoryMetrics contains guest memory statistics for the balloon controller.
```
Comment (AkihiroSuda) [inline pkg/guestagent/api/guestservice.proto:60]: https://pkg.go.dev/github.com/containerd/cgroups/v3@v3.1.3/cgroup2/stats#MemoryStat should suffice?

File: cmd/lima-guestagent/daemon_linux.go [hunk 2]
```diff
@@ -145,5 +146,8 @@ func daemonAction(cmd *cobra.Command, _ []string) error {
 		logrus.Infof("serving the guest agent on %q", socket)
 	}
 	defer logrus.Debug("exiting lima-guestagent daemon")
-	return server.StartServer(ctx, l, &server.GuestServer{Agent: agent, TunnelS: portfwdserver.NewTunnelServer()})
+	dockerSocket := "/var/run/docker.sock"
```
Comment (AkihiroSuda) [inline cmd/lima-guestagent/daemon_linux.go:149]: - The path is different for rootless Docker and Podman
- Why only support containerized processes?

File: hack/bats/tests/alpine-docker-vz.bats
Comment (AkihiroSuda) [inline hack/bats/tests/alpine-docker-vz.bats:1]: Most tests seem irrelevant to memory optimization

File: pkg/guestagent/metrics/collector.go
Comment (AkihiroSuda) [inline pkg/guestagent/metrics/collector.go:1]: Linux specific codes must have `_linux.go` suffix

File: pkg/guestagent/metrics/collector.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+	m.ContainerCpuPercent = dockerCPU
+	m.ContainerIoBytesPerSec = dockerIO
+
+	return m, nil
+}
+
+// collectDockerStats queries the Docker socket for container count,
+// aggregate CPU%, and aggregate IO bytes/sec. Containers are polled
+// in parallel with a 3-second overall timeout. Returns zeros on error.
+func (c *Collector) collectDockerStats(ctx context.Context) (count int, cpuPercent, ioBytesPerSec float64) {
```
Comment (AkihiroSuda) [inline pkg/guestagent/metrics/collector.go:178]: Why depend on Docker API? Why not directly stat the cgroupfs?

File: pkg/limatype/lima_instance.go [hunk 4]
```diff
@@ -32,8 +33,9 @@ type Instance struct {
 	VMType                VMType            `json:"vmType"`
 	Arch                  Arch              `json:"arch"`
 	CPUs                  int               `json:"cpus,omitempty"`
-	Memory                int64             `json:"memory,omitempty"` // bytes
-	Disk                  int64             `json:"disk,omitempty"`   // bytes
+	Memory                int64             `json:"memory,omitempty"`         // bytes (configured)
+	PhysicalMemory        int64             `json:"physicalMemory,omitempty"` // bytes (actual host footprint)
```
Comment (AkihiroSuda) [inline pkg/limatype/lima_instance.go:37]:
```suggestion
	PhysicalMemory       *int64             `json:"physicalMemory,omitempty"` // bytes (actual host footprint)
```
As this information will be only available for specific VM drivers

File: pkg/limatype/lima_yaml.go [hunk 5]
```diff
@@ -121,8 +121,85 @@ type QEMUOpts struct {
...
+	SettleWindow *string `yaml:"settleWindow,omitempty" json:"settleWindow,omitempty" jsonschema:"nullable"`
+	// MaxContainerCPU is the container CPU usage threshold that blocks shrinking (percentage, e.g., 10.0).
+	MaxContainerCPU *float64 `yaml:"maxContainerCPU,omitempty" json:"maxContainerCPU,omitempty" jsonschema:"nullable"`
+	// MaxContainerIO is the container I/O rate threshold that blocks shrinking (e.g., "100MiB").
+	MaxContainerIO *string `yaml:"maxContainerIO,omitempty" json:"maxContainerIO,omitempty" jsonschema:"nullable"`
+	// FloorStaleness is how long a learned floor stays valid before re-learning (e.g., "24h"). 0 = never stale.
+	FloorStaleness *string `yaml:"floorStaleness,omitempty" json:"floorStaleness,omitempty" jsonschema:"nullable"`
+	// EnableTrendDetection enables pre-emptive grow on rising PSI trend (avg10 > 1.5*avg60).
+	EnableTrendDetection *bool `yaml:"enableTrendDetection,omitempty" json:"enableTrendDetection,omitempty" jsonschema:"nullable"`
+}
```
Comment (AkihiroSuda) [inline pkg/limatype/lima_yaml.go:171]: Do you really need all these configuration knobs?

File: pkg/store/physmem_darwin.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+		logrus.Debugf("Could not get physical footprint for PID %d: %v", pid, err)
+		return 0
+	}
+	return mem
+}
+
+// findDiskOwnerPID finds the PID of the process that has the given disk file open
+// using the macOS `fuser` command.
+func findDiskOwnerPID(diskPath string) (int, error) {
+	out, err := exec.CommandContext(context.Background(), "fuser", diskPath).CombinedOutput()
```
Comment (AkihiroSuda) [inline pkg/store/physmem_darwin.go:39]: Needs `LANG=C` `LC_ALL=C`?
Same for other commands too

Ideally this should rather call `PROC_PIDLISTFDS`

File: templates/experimental/alpine-docker-vz.yaml
Comment (AkihiroSuda) [inline templates/experimental/alpine-docker-vz.yaml:1]: Seems too complicated.
I suggest moving it to a third-party repo: https://lima-vm.io/docs/templates/github/
e.g. `github:nixos-lima` https://github.com/nixos-lima/nixos-lima

File: website/content/en/docs/config/vmtype/vz.md [hunk 7]
```diff
@@ -27,6 +27,178 @@ base:
 ```
 {{% /tab %}}
 {{< /tabpane >}}
+
+### Memory Ballooning
+
+| ⚡ Requirement | Lima >= 2.0.0, macOS >= 13.0, VZ backend only |
```
Comment (AkihiroSuda) [inline website/content/en/docs/config/vmtype/vz.md:33]: Not in v2.0

**Comment by jwehrlich** [comment]
> This amount of PR is not reviewable.
> 
> Please split the PR into multiple ones. Notably auto-pause stuff does not seem directly related to memory optimization and does not need to be overloaded in this PR?

**Comment by thewesjohnson** [comment]
@jwehrlich — I tested balloon reclamation on macOS 26.0.1 (M1 Max, 32GB, Colima 0.10.1, vz backend) and can confirm
  the host-side issue your controller is working around.

My test captures both sides: guest /proc/meminfo shows 108% memory recovery after free + drop_caches, but the host
  com.apple.Virtualization.VirtualMachine process RSS never decreases — it grew from 1731 to 1748 MB during a 120-second observation window with zero decreases.

Your adaptive controller's guest-side balloon inflation works correctly — the problem is entirely at Apple's layer. Once they fix the host-side madvise, your controller should just work.

Full test data and reproduction script: https://github.com/thewesjohnson/macos-virtio-balloon-test

Happy to run your branch against the same workload if that would help validate your state transitions (Bootstrap -> LearningDescend -> Steady -> OOMRecovery).

I've filed Apple Feedback **FB22614752** with the full dataset.

**Comment by nagstaku** [comment]
**Disclosure: this comment was researched and written by an AI agent (Claude) working with me on my own machine.** If that breaches a norm here, apologies — tell me and I'll remove it. Posting purely as information, **not** as a recommendation or a request for anyone to act on.

@jwehrlich — I built this branch to test whether VZ ballooning can return memory to macOS, and wrote the results up in https://github.com/lima-vm/lima/issues/4220#issuecomment-5142393637 rather than duplicating them here.

Short version, on macOS 26.5.2 / Apple Silicon / 16 GB: your controller's **guest side works exactly as designed**. I patched in a manual target override to isolate it from the adaptive state machine, and driving the balloon to a 1 GiB floor inflated precisely 7 GiB (`3145728 - 1310720 = 1835008` pages x 4KiB) with the guest dropping to 325 MB free. Textbook.

`Virtualization.framework` then released none of it — host footprint sat at 8010 MB for 7 minutes, including with `compact_memory` before and during, per Apple's own header advice. Reading the negotiated feature bitmap, Apple's balloon never offers `VIRTIO_BALLOON_F_REPORTING` (bit 5), which is the bit that drives host-side reclaim; the guest kernel supports it, the device just doesn't advertise it.

I want to be careful about what that does and doesn't mean, since it's your work and only my machine: it isn't a defect in this PR, and it doesn't mean the approach is wrong. It may simply mean the payoff is gated on something outside Lima. Whether that makes the split-up work worth doing now or worth parking until Apple changes the device is entirely yours and the maintainers' call — I only thought you'd rather have the measurement before investing more effort in splitting 4,000+ lines.

Happy to re-run anything, or to test a revision of the controller if that's useful.
